external-crates/move/crates/move-compiler/src/parser/syntax.rs

```diff
@@ -1422,21 +1480,188 @@ fn parse_name_exp(context: &mut Context) -> Result<Exp_, Box<Diagnostic>> {
+                let pat = parse_match_pattern(context);
+                consume_token(context.tokens, Tok::RParen)?;
+                pat
+            }
+            Tok::Identifier if context.tokens.content() == "_" => {
+                context.tokens.advance()?;
+                Ok(sp(
+                    context.tokens.current_token_loc(),
+                    MatchPattern_::Wildcard,
+                ))
+            }
+            Tok::Identifier => ok_with_loc!(context, {
+                let name_access_chain = parse_name_access_chain(context, || "a pattern entry")?;
+                let ty_args = parse_optional_type_args(context)?;
+
+                match context.tokens.peek() {
+                    Tok::LParen => {
+                        let (loc, patterns) = with_loc!(
+                            context,
+                            parse_comma_list(
+                                context,
+                                Tok::LParen,
+                                Tok::RParen,
+                                parse_match_pattern,
+                                "a pattern",
+                            )?
+                        );
+                        PositionalConstructor(name_access_chain, ty_args, sp(loc, patterns))
+                    }
+                    Tok::LBrace => {
+                        let (loc, patterns) = with_loc!(
+                            context,
+                            parse_comma_list(
+                                context,
+                                Tok::LBrace,
+                                Tok::RBrace,
+                                parse_field_pattern,
+                                "a field pattern",
+                            )?
+                        );
+                        FieldConstructor(name_access_chain, ty_args, sp(loc, patterns))
+                    }
+                    _ => Name(name_access_chain, ty_args),
+                }
+            }),
+            _ => {
+                if let Some(value) = maybe_parse_value(context)? {
+                    Ok(sp(value.loc, Literal(value)))
+                } else {
+                    Err(Box::new(diag!(
+                        Syntax::UnexpectedToken,
+                        (context.tokens.current_token_loc(), INVALID_PAT_ERROR_MSG)
+                    )))
+                }
+            }
+        }
+    }
+
+    fn parse_field_pattern(
+        context: &mut Context,
+    ) -> Result<(Field, MatchPattern), Box<Diagnostic>> {
+        let field = parse_field(context)?;
+        let pattern = if match_token(context.tokens, Tok::Colon)? {
+            parse_match_pattern(context)?
+        } else {
+            sp(
+                field.loc(),
+                Name(sp(field.loc(), NameAccessChain_::One(field.0)), None),
+            )
+        };
+        Ok((field, pattern))
+    }
+
+    fn parse_at_pattern(context: &mut Context) -> Result<MatchPattern, Box<Diagnostic>> {
```

Comment:
Nit: this naming doesn't seem quite right since this is parsing more than just at patterns. Maybe something like `parse_top_level_pattern`/`parse_pattern` or something like that?

Reply:
Revised this to help explain what is going on, plus fix a bug